core/Pos/Core/Genesis/Types.hs

```diff
@@ -57,20 +48,44 @@ getTotalStake ExponentialStakes = mkCoin . sum $
     let g 0 = []
         g n = n : g (n `div` 2)
     in g 5000
-getTotalStake (ExplicitStakes balances) = unsafeIntegerToCoin $
-    sumCoins $ fst <$> balances
-getTotalStake (CombinedStakes st1 st2) =
-    getTotalStake st1 `unsafeAddCoin` getTotalStake st2
+getTotalStake (CustomStakes balances) =
+    unsafeIntegerToCoin $ sumCoins balances
 
--- | Hardcoded genesis data
-data GenesisCoreData = GenesisCoreData
-    { gcdAddresses         :: [Address]
-    , gcdDistribution      :: StakeDistribution
-    , gcdBootstrapBalances :: !(HashMap StakeholderId Coin)
+-- | List of distributions accompained by related addresses set (what
+-- to distribute and how).
+type AddrDistribution = [(HashSet Address, StakeDistribution)]
+
+-- | Hardcoded genesis data to generate utxo from.
+data GenesisCoreData = UnsafeGenesisCoreData
+    { gcdAddrDistribution      :: !AddrDistribution
+      -- ^ Address distribution. Determines utxo without boot
+      -- stakeholders distribution (addresses and coins).
+    , gcdBootstrapStakeholders :: !(HashSet StakeholderId)
+      -- ^ Bootstrap era stakeholders.
     }
     deriving (Show, Eq)
 
-instance Monoid GenesisCoreData where
-    mempty = GenesisCoreData mempty mempty mempty
-    (GenesisCoreData addrsA distA bbsA) `mappend` (GenesisCoreData addrsB distB bbsB) =
-        GenesisCoreData (addrsA <> addrsB) (distA <> distB) (bbsA <> bbsB)
+-- | Safe constructor for 'GenesisCoreData'. Throws error if something is wrong.
+mkGenesisCoreData ::
+       [(HashSet Address, StakeDistribution)]
+    -> HashSet StakeholderId
+    -> Either String GenesisCoreData
+mkGenesisCoreData distribution bootStakeholders = do
+    -- TODO add checks
+    -- 1. Every set of address matches by the size to distribution size
+    -- 2. If using CustomStakes, lengths match
+    -- 3. (?) Addresses in (map fst) are unique
+    -- 4. All the stake is distributed to gcdBootstrapStakeholders
+    pure $ UnsafeGenesisCoreData distribution bootStakeholders
+
+-- | Concats 'GenesisCoreData'. We do not use 'Monoid' instance for
+-- 'GenesisCoreData' because 'mempty' doesn't make sense (empty boot
+-- stakeholders?).
```

Comment:
You could use `Semigroup`, but apparently you can't do it, because it's not guaranteed that `concatGenesisCoreData validGenesisCoreData1 validGenesisCoreData2` is valid. I didn't try to understand it, but if it's guaranteed, then I don't see a reason why this function returns `Either`.